### PR TITLE
feat: add allow list to `no-template-shadow`

### PR DESCRIPTION
**Pull request by mussinbenarbia**
This is an attempt at solving https://github.com/vuejs/eslint-plugin-vue/issues/2317

As suggested by @FloEdelmann, I tried to match the `no-shadow` core ESLint rule and called the option `allow`.
I have not modified the docs yet as I'd like to make sure the implementation is OK first.

### CODE REVIEW

File: lib/rules/no-template-shadow.js [hunk 1]
```diff
@@ -30,7 +37,19 @@ module.exports = {
...
+        type: 'object',
+        properties: {
+          allow: {
+            type: 'array',
+            items: {
+              type: 'string'
+            }
+          }
+        }
+      }
```
Comment (FloEdelmann) [inline lib/rules/no-template-shadow.js:51]:
```suggestion
      {
        type: 'object',
        properties: {
          allow: {
            type: 'array',
            items: {
              type: 'string'
            },
            uniqueItems: true
          }
        },
        additionalProperties: false
      }
```

**Review by FloEdelmann — COMMENTED** [review]
Looks good to me, thanks :slightly_smiling_face: 
Apart from one small suggestion and the missing documentation, this is ready to go.

**Comment by mussinbenarbia** [comment]
@FloEdelmann Thanks for the suggestion! I made the change and updated the docs. Please take a look when you get the chance 🙏

File: docs/rules/no-template-shadow.md [hunk 2]
```diff
@@ -36,21 +36,51 @@ This rule aims to eliminate shadowed variable declarations of v-for directives o
...
+}
 </script>
 ```
 
 </eslint-code-block>
 
 ## :wrench: Options
 
-Nothing.
+This rule takes one option, an object, with the property `"allow"`.
```
Comment (FloEdelmann) [inline docs/rules/no-template-shadow.md:53]:
```suggestion
This rule takes one optional object option, with the property `"allow"`.
```

**Review by FloEdelmann — APPROVED** [review]
One minor nit, otherwise this looks good to me :slightly_smiling_face:

File: docs/rules/no-template-shadow.md (hunk 2 above)
Comment (mussinbenarbia) [reply docs/rules/no-template-shadow.md]: Thanks! Done 👍

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**